blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/MetricIndexData.java

```diff
@@ -0,0 +1,204 @@
+package com.rackspacecloud.blueflood.io;
+
+import java.util.*;
+
+/**
+ * When we ingest these below metrics, ES generates the following indexes
+ *
+ *     metric           indexes
+ *   -----------        -------
+ *   foo.bar.baz.aux -> [foo, foo.bar, foo.bar.baz, foo.bar.baz.aux, bar, baz, aux]
+ *   foo.bar.baz     -> [foo, foo.bar, foo.bar.baz, bar, baz]
+ *
+ * If we request ES to return us upto next two levels of the metric indexes for foo.bar,
+ * we get this below response from ES where key is the metric index and doc_count is
+ * the number of documents(think metric names) it points to.
+ *
+ * "buckets" : [ {
+ *     "key" : "foo.bar",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz",
+ *     "doc_count" : 2
+ *   }, {
+ *     "key" : "foo.bar.baz.aux",
+ *     "doc_count" : 1
+ * } ]
+ *
+ * If we feed each metric index to this class by calling the add(metricIndex, docCount) method,
+ * this class analyzes the data and provides methods to retrieve different aspects of this data
+ * with respect to baseLevel(if set to 2 with above data, uses foo.bar as reference).
+ *
+ */
+public class MetricIndexData {
+
+    private static class MetricIndexDocCount {
+
+        private long actualDocCount;
+        private long childrenTotalDocCount;
+
+        public MetricIndexDocCount(long actualDocCount, long childrenTotalDocCount) {
+            this.actualDocCount = actualDocCount;
+            this.childrenTotalDocCount = childrenTotalDocCount;
+        }
+
+        public void setActualDocCount(long actualDocCount) {
+            this.actualDocCount = actualDocCount;
+        }
+
+        public void addChildrenTotalDocCount(long childrenDocCount) {
+            this.childrenTotalDocCount += childrenDocCount;
+        }
+    }
+
+    private final int baseLevel;
+
+    //contains next level of tokens after baseLevel, which also have subsequent next level
+    private final Set<String> tokensWithNextLevelSet = new LinkedHashSet<String>();
+
+    //contains all metric indexes which are of the same length as baseLevel
+    final Map<String, MetricIndexDocCount> metricIndexSameLevelMap = new HashMap<String, MetricIndexDocCount>();
+
+    //contains all metric indexes which are more than the length of baseLevel by one.
+    final Map<String, MetricIndexDocCount> metricIndexNextLevelMap = new HashMap<String, MetricIndexDocCount>();
+
+    public MetricIndexData(int baseLevel) {
+        this.baseLevel = baseLevel;
+    }
+
+    /**
+     * For a given metricIndex and docCount, classifies the data with respect to baseLevel
+     * and stores it accordingly.
+     *
+     * @param metricIndex
+     * @param docCount is document count
+     */
+    public void add(String metricIndex, long docCount) {
```

Comment:
The name metricIndex confused me briefly, until I see how this method is being called. It's really a metric name. Maybe a simple comment (in the @param metricIndex) would help?